src/components/swap/ConfirmSwapModal.tsx

```diff
@@ -1,39 +1,172 @@
 import SwapModalFooter from './SwapModalFooter'
 import SwapModalHeader from './SwapModalHeader'
 
+export enum ConfirmModalState {
+  REVIEWING,
+  APPROVING_TOKEN,
+  PERMITTING,
+  PENDING_CONFIRMATION,
+}
+
+function isAllowing(confirmModalState: ConfirmModalState) {
+  return confirmModalState === ConfirmModalState.APPROVING_TOKEN || confirmModalState === ConfirmModalState.PERMITTING
+}
+
+function useConfirmModalState({
+  trade,
+  allowedSlippage,
+  onSwap,
+  allowance,
+  doesTradeDiffer,
+}: {
+  trade: InterfaceTrade
+  allowedSlippage: Percent
+  onSwap: () => void
+  allowance: Allowance
+  doesTradeDiffer: boolean
+}) {
+  const [confirmModalState, setConfirmModalState] = useState<ConfirmModalState>(ConfirmModalState.REVIEWING)
+  const [approvalError, setApprovalError] = useState<PendingModalError>()
+  const [pendingModalSteps, setPendingModalSteps] = useState<PendingConfirmModalState[]>([])
+
+  const prepareSwapFlow = useCallback(() => {
+    const steps: PendingConfirmModalState[] = []
+    if (allowance.state === AllowanceState.REQUIRED && allowance.needsPermit2Approval) {
+      steps.push(ConfirmModalState.APPROVING_TOKEN)
+    }
+    if (allowance.state === AllowanceState.REQUIRED && allowance.needsSignature) {
+      steps.push(ConfirmModalState.PERMITTING)
+    }
+    steps.push(ConfirmModalState.PENDING_CONFIRMATION)
+    setPendingModalSteps(steps)
+  }, [allowance])
+
+  const { chainId } = useWeb3React()
+  const trace = useTrace()
+  const maximumAmountIn = useMaxAmountIn(trade, allowedSlippage)
+  // Starts the approval process, by triggering either the Token Approval or the Permit signature.
+  const updateAllowance = useCallback(async () => {
+    invariant(allowance.state === AllowanceState.REQUIRED)
+    try {
+      if (allowance.needsPermit2Approval) {
+        setConfirmModalState(ConfirmModalState.APPROVING_TOKEN)
+        await allowance.approve()
+        sendAnalyticsEvent(InterfaceEventName.APPROVE_TOKEN_TXN_SUBMITTED, {
+          chain_id: chainId,
+          token_symbol: maximumAmountIn?.currency.symbol,
+          token_address: maximumAmountIn?.currency.address,
+          ...trace,
+        })
+      } else {
+        setConfirmModalState(ConfirmModalState.PERMITTING)
+        await allowance.permit()
+      }
+    } catch (e) {
+      setConfirmModalState(ConfirmModalState.REVIEWING)
+      if (didUserReject(e)) {
+        return
+      }
+      console.error(e)
+      setApprovalError(
+        allowance.needsPermit2Approval ? PendingModalError.TOKEN_APPROVAL_ERROR : PendingModalError.PERMIT_ERROR
+      )
+    }
+  }, [allowance, chainId, maximumAmountIn?.currency.address, maximumAmountIn?.currency.symbol, trace])
+
+  const onStartSwapFlow = useCallback(async () => {
+    setApprovalError(undefined)
+    if (allowance.state === AllowanceState.REQUIRED) {
+      await updateAllowance()
+    } else {
+      setConfirmModalState(ConfirmModalState.PENDING_CONFIRMATION)
+      onSwap()
```

Comment:
I think instead of calling `await updateAllowance()`, you should just do `await allowance.approve()` and just move the body of that function here instead.

My point is that `updateAllowance` sets some states for the component as well, so it immediately becomes more layered to understand how all state updates are performed here. Ideally, all state updates were performed in this function and only in this function and other functions would be simply for side effects.

Reply:
yeah, i can move this logic into `startSwapFlow` and just call `startSwapFlow` from the effect hooks below